File: csharp/TargetItem.cs

namespace csharp
{
    public class TargetItem : ITarget
    {
        public Item Item { get; set; }

        public TargetItem(Item item)
        {
            this.Item = item;
        }
        public virtual void CalculateSellByDate()
        {
            this.Item.SellIn -= 1;
        }
        public virtual void CalculateQuality()
        {
            Item.Quality--;
            if (Item.SellIn < 0)
            {
                Item.Quality--;
            }
        }
        public virtual void SpecifyQualityValue()
        {
            if (Item.Quality < 0)
            {
                Item.Quality = 0;
            }
            else if (Item.Quality > 50)
            {
                Item.Quality = 50;
            }
        }
    }
}

File: TypeScript/app/target-item.ts
import { Item } from "./gilded-rose";

export interface ITarget {
    item: Item;
    calculateSellByDate(): void;
    calculateQuality(): void;
    specifyQualityValue(): void;
}

export class TargetItem implements ITarget {
    item: Item;
    constructor(item: Item) {
        this.item = item;
    }

    calculateSellByDate(): void {
        this.item.sellIn -= 1;
    }
    calculateQuality(): void {
        this.item.quality--;
        if (this.item.sellIn < 0) {
            this.item.quality--;
        }
    }
    specifyQualityValue(): void {
        if (this.item.quality < 0) {
            this.item.quality = 0;
        }
        else if (this.item.quality > 50) {
            this.item.quality = 50;
        }
    }
}
